zapcore/clock_test.go

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zapcore
+
+import (
+	"testing"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/atomic"
+	"go.uber.org/goleak"
+)
+
+type controlledClock struct{ *clock.Mock }
+
+func newControlledClock() *controlledClock {
+	return &controlledClock{clock.NewMock()}
+}
+func (c *controlledClock) NewTicker(d time.Duration) *time.Ticker {
+	return &time.Ticker{C: c.Ticker(d).C}
+}
+
+func TestMockClock(t *testing.T) {
+	defer goleak.VerifyNone(t)
+	var n atomic.Int32
+	ctrlMock := newControlledClock()
+
+	quit := make(chan bool)
+	// Create a channel to increment every microsecond.
+	go func(ticker *time.Ticker) {
+		for {
+			select {
+			case <-quit:
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				n.Inc()
+			}
+		}
+	}(ctrlMock.NewTicker(time.Microsecond))
+
+	// Move clock forward.
+	ctrlMock.Add(2 * time.Microsecond)
+	assert.Equal(t, int32(2), n.Load())
+	quit <- true
```

Comment:
For channels that'll only ever fire once, `close` the channel instead of posting to it. (See also [this](https://abhinavg.net/posts/fire-once-antipattern/).)

Reply:
Separately, please also see [this](https://github.com/uber-go/zap/pull/948#discussion_r637262837) comment.

We need to wait for the goroutine to actually exit once we tell it to stop before the test finishes.

Consider incorporating [goleak](https://github.com/uber-go/goleak) with a TestMain in zapcore to detect issues like this.